### PR TITLE
Adding logic to handle JWT when logging out

### PR DESCRIPTION
**Pull request by AnhChienVu**
Fixes: #18 

## Feature:
Adding a log out button that allows user to log out.

## Issue's Requirements:
- **Frontend**:
  - Call all a /logout api endpoint: I created a Logout component under `frontend/src/components/logout/index.jsx`. 
  - Clear the jwt from the local storage: when user click to that button, it will remove the jwt from the local storage as well as make a POST request to the '/api/logout' from the backend side along with the `Authentication Header` that will be the JWT token.
  -  Redirect the user to the login (or just refresh the page):  I cannot use the location to fulfill it so I used `window.location.href = "/";` instead to redirect to '/' page
- **Backend**:
   - On a call to the /logout api endpoint add the jwt id to a list of invalidated tokens: As i see in your existing function `generateAuthToken()` at `backend/auth/utils.go` file didn't generate the `ID` property for the JWT token, so after logging in, the parsed JWT token didn't have that value, so I had to add it to fulfill your task of adding JWT id to invalidated list
   - In the jwt auth middleware, reject tokens that are in this list with 403: I added a function to handle Log out request in `backend/auth/handlers.go`. Inside this function I want to add the JWT id to the invalidated list that was defined in `backend/auth/jwt_middleware.go`. 
      Because the `CheckAuth` middleware is called in every request so had an if statement to check if the JWT id extracted from a JWT tokens was already existed in the invalidated list, if it is, then don't let user log in.

## Manual Testing
After logging to the app, I copied the JWT Token from local storage, then click log out button. I manually input the JWT token to local storage to check if I can log in or not. Because after log out the JWT id is unique and it will be added to the list of invalidated tokens. And if user want to use the same id within a token to access then it will fail.

### CODE REVIEW

File: backend/auth/handlers.go [hunk 1]
```diff
@@ -1,118 +1,157 @@
...
+
+    c.JSON(http.StatusOK, gin.H{
+        "token": signedToken,
+    })
 }
+
+// POST /logout
+// Logout invalidates the current JWT token
+func (h *AuthHandlers) Logout(c *gin.Context) {
+    authHeader := c.GetHeader("Authorization")
```
Comment (zer0tonin) [inline backend/auth/handlers.go:128]: For the sake of being DRY we can do this in JWTMiddleware.CheckAuth. I think it should just set the jti in the context

File: frontend/src/components/logout/index.jsx [hunk 2]
```diff
@@ -0,0 +1,41 @@
+import { h } from "preact";
+
+import { useLocation } from "preact-iso";
+import "./style.css";
+
+const Logout = () => {
+  const location = useLocation();
+  const handleLogout = async () => {
+    try {
+      const token = localStorage.getItem("jwt");
```
Comment (zer0tonin) [inline frontend/src/components/logout/index.jsx:10]: Please use the jwt context like we do in the rest of the code:
```
const { jwt } = useContext(AuthContext);
```

File: frontend/src/components/logout/index.jsx [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+      }
+      await fetch("/api/logout", {
+        method: "POST",
+        headers: {
+          "Content-Type": "application/json",
+          Authorization: `Bearer ${token}`,
+        },
+      });
+
+      localStorage.removeItem("jwt");
```
Comment (zer0tonin) [inline frontend/src/components/logout/index.jsx:23]: This part should be in a `finally` block. The frontend should remove the jwt regardless of the success of the /logout call

File: frontend/src/components/logout/index.jsx [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+      //   location.route("/");
+      window.location.href = "/";
```
Comment (zer0tonin) [inline frontend/src/components/logout/index.jsx:26]: Can you use the preact version:

```
const location = useLocation()
location.route("/")
```

**Review by zer0tonin — CHANGES_REQUESTED** [review]
Thank you for this PR.

I think it looks good so far, I added a few request just to make sure the code stays consistent with the rest of the codebase.

Would you mind also running `go fmt` on your go code? I think your tab settings are not what's used in the rest of the project.

File: frontend/src/components/logout/index.jsx (hunk 4 above)
Comment (AnhChienVu) [reply frontend/src/components/logout/index.jsx]: Done

File: frontend/src/components/logout/index.jsx (hunk 3 above)
Comment (AnhChienVu) [reply frontend/src/components/logout/index.jsx]: Done

File: frontend/src/components/logout/index.jsx (hunk 2 above)
Comment (AnhChienVu) [reply frontend/src/components/logout/index.jsx]: Done

File: backend/auth/handlers.go (hunk 1 above)
Comment (AnhChienVu) [reply backend/auth/handlers.go:128]: Done

**Comment by AnhChienVu** [comment]
I just made some changes that satisfy your request

**Pull request merged**